server/src/main/java/org/elasticsearch/indices/IndicesService.java

```diff
@@ -587,28 +595,28 @@ public CompletableFuture<IndexShard> createShard(ClusterState state,
         ShardId shardId = shardRouting.shardId();
         CompletableFuture<IndexShard> future = new CompletableFuture<>();
         CompletableFuture<IndexShard> pending = pendingShardCreations.putIfAbsent(shardId, future);
-        if (pending == null) {
```

Comment:
What are the reasons for the changes here?

If it now runs into the race where `pending` is returned, it won't execute `pendingShardCreations.remove(shardId)` or the `startRecovery`. Is that intentional?

Reply:
The intention is the opposite as before, `startRecovery` was called multiple times, first when creating the future, and then each time the pending was returned, resulting in an exception while running the tests with some cluster state changes arriving.
Why do you think that the attached calls aren't executed anymore when returning the future from the pending map?

Reply:
Ah, got it.

> Why do you think that the attached calls aren't executed anymore when returning the future from the pending map?

I misread the diff